File: compile_csproj.py
import argparse
import concurrent.futures
import dataclasses
import json
import os
import subprocess
import xml.etree.ElementTree as ET
from collections.abc import Iterable
from concurrent.futures import ThreadPoolExecutor
from pathlib import Path
from typing import Any
from xml.dom import minidom

BAZEL_BINARY = "bazel"


def _run_paket2bazel(directory: Path):
    subprocess.run(
        [
            BAZEL_BINARY,
            "run",
            "@rules_dotnet//tools/paket2bazel",
            "--",
            "--dependencies-file",
            str(directory / "paket.dependencies"),
            "--output-folder",
            str(directory),
        ],
        capture_output=True,
        text=True,
        cwd=directory,
    )


def _run_bazel_build(directory: Path):
    subprocess.run(
        [
            BAZEL_BINARY,
            "build",
            "//...",
        ],
        capture_output=True,
        text=True,
        cwd=directory,
    )


def _run_bazel_aquery(directory: Path):
    process_result = subprocess.run(
        [BAZEL_BINARY, "aquery", "//...", "--output=jsonproto"],
        check=True,
        capture_output=True,
        cwd=directory,
    )
    aquery = json.loads(process_result.stdout)
    return aquery


def _run_bazel_output_base(directory: Path):
    process_result = subprocess.run(
        [BAZEL_BINARY, "info", "output_base"],
        check=True,
        capture_output=True,
        cwd=directory,
        text=True,
    )
    output_base = process_result.stdout.strip()
    return Path(output_base)


def _run_bazel_output_path(directory: Path):
    process_result = subprocess.run(
        [BAZEL_BINARY, "info", "output_path"],
        check=True,
        capture_output=True,
        cwd=directory,
        text=True,
    )
    output_path = process_result.stdout.strip()
    return Path(output_path)


def _run_bazel_workspace(directory: Path):
    process_result = subprocess.run(
        [BAZEL_BINARY, "info", "workspace"],
        check=True,
        capture_output=True,
        cwd=directory,
        text=True,
    )
    workspace = process_result.stdout.strip()
    return Path(workspace)


def _list_symbols(list_symbols_binary: Path, file: Path):
    result = subprocess.run(
        list(map(str, [list_symbols_binary, file])), capture_output=True, text=True
    )
    return result.stdout


@dataclasses.dataclass
class _Action:
    target_id: int
    inputs: list[Path]
    outputs: list[Path]
    primary_output: Path


def _get_path(path_fragments: list[dict], path_fragment_id: int):
    fragment = (fragment,) = [
        other_fragment
        for other_fragment in path_fragments
        if other_fragment["id"] == path_fragment_id
    ]
    related_fragments: list[str] = [fragment["label"]]
    while "parentId" in fragment:
        (fragment,) = [
            other_fragment
            for other_fragment in path_fragments
            if other_fragment["id"] == fragment["parentId"]
        ]
        related_fragments.append(fragment["label"])

    path = Path(related_fragments.pop())
    while related_fragments:
        path = path / related_fragments.pop()
    return path


def _select(iterable: Iterable[dict[str, Any]], key: str, value: int) -> dict[str, Any]:
    (item,) = [item for item in iterable if item[key] == value]
    return item


class _ActionsFactory:
    def __init__(self, aquery: dict[str, Any]):
        self.aquery = aquery

    def __call__(self) -> list[_Action]:
        actions: list[_Action] = []
        for aquery_action in self.aquery["actions"]:
            if aquery_action["mnemonic"] == "CSharpCompile":
                action = self._get_action(aquery_action["targetId"])
                actions.append(action)
        return actions

    def _get_artifact_path(self, artifact_id: int):
        artifact = _select(self.aquery["artifacts"], "id", artifact_id)
        path_fragment_id = artifact["pathFragmentId"]
        path = _get_path(self.aquery["pathFragments"], path_fragment_id)
        return path

    def _recursively_get_inputs(self, dep_set_id: int) -> list[Path]:
        inputs = []
        dep_set_of_files = _select(self.aquery["depSetOfFiles"], "id", dep_set_id)
        direct_artifact_ids = dep_set_of_files["directArtifactIds"]
        for direct_artifact_id in direct_artifact_ids:
            path = self._get_artifact_path(direct_artifact_id)
            inputs.append(path)
        for transitive_dep_set_id in dep_set_of_files.get("transitiveDepSetIds", []):
            inputs.extend(self._recursively_get_inputs(transitive_dep_set_id))
        return inputs

    def _get_action(self, target_id: int) -> _Action:
        inputs: list[Path] = []
        outputs: list[Path] = []
        (action,) = [
            action
            for action in self.aquery["actions"]
            # A targetId can correspond to multiple actions:
            if action["targetId"] == target_id and action["mnemonic"] == "CSharpCompile"
        ]
        input_dep_set_ids = action["inputDepSetIds"]
        for input_dep_set_id in input_dep_set_ids:
            dep_set_of_files = _select(
                self.aquery["depSetOfFiles"], "id", input_dep_set_id
            )

            direct_artifact_ids = dep_set_of_files["directArtifactIds"]
            for direct_artifact_id in direct_artifact_ids:
                path = self._get_artifact_path(direct_artifact_id)
                inputs.append(path)

            transitive_dep_set_ids = dep_set_of_files.get("transitiveDepSetIds", [])
            for transitive_dep_set_id in transitive_dep_set_ids:
                paths = self._recursively_get_inputs(transitive_dep_set_id)
                inputs.extend(paths)

        output_ids = action["outputIds"]
        for output_id in output_ids:
            path = self._get_artifact_path(output_id)
            outputs.append(path)

        primary_output_id = action["primaryOutputId"]
        primary_output_artifact = _select(
            self.aquery["artifacts"], "id", primary_output_id
        )
        primary_output_path_fragment_id = primary_output_artifact["pathFragmentId"]
        primary_output_path = _get_path(
            self.aquery["pathFragments"], primary_output_path_fragment_id
        )

        (dotnet_cli_home,) = [
            env_var["value"]
            for env_var in action["environmentVariables"]
            if env_var["key"] == "DOTNET_CLI_HOME"
        ]
        dotnet_cli_home_path = Path(dotnet_cli_home)
        inputs = [
            input_path
            for input_path in inputs
            if not input_path.is_relative_to(dotnet_cli_home_path)
        ]
        return _Action(
            target_id=target_id,
            inputs=inputs,
            outputs=outputs,
            primary_output=primary_output_path,
        )


def _get_actions(directory: Path) -> list[_Action]:
    _run_paket2bazel(directory)
    _run_bazel_build(directory)
    aquery = _run_bazel_aquery(directory)
    actions_factory = _ActionsFactory(aquery)
    return actions_factory()


def _generate_csproj(
    csproj_path: Path,
    source_files: list[str],
    dll_references: list[str],
    target_framework: str,
) -> None:
    project = ET.Element("Project", Sdk="Microsoft.NET.Sdk")
    prop_group = ET.SubElement(project, "PropertyGroup")
    tf = ET.SubElement(prop_group, "TargetFramework")
    tf.text = target_framework

    item_group_sources = ET.SubElement(project, "ItemGroup")
    for src in source_files:
        ET.SubElement(item_group_sources, "Compile", Include=src)

    item_group_refs = ET.SubElement(project, "ItemGroup")
    for dll in dll_references:
        ref = ET.SubElement(
            item_group_refs, "Reference", Include=dll.split("/")[-1].replace(".dll", "")
        )
        hint = ET.SubElement(ref, "HintPath")
        hint.text = dll

    xmlstr = minidom.parseString(ET.tostring(project)).toprettyxml(indent="  ")
    csproj_path.write_text(xmlstr)


def _compile_csproj(
    project_name: str, directory: Path, target_framework: str, list_symbol_binary: Path
) -> None:
    if not list_symbol_binary.exists():
        raise FileNotFoundError(f"ListSymbols binary not found: {list_symbol_binary}")

    workspace_absolute = Path(os.environ["BUILD_WORKSPACE_DIRECTORY"])

    directory = workspace_absolute / directory

    actions = _get_actions(directory)

    output_base = _run_bazel_output_base(directory)

    bazel_output_path = _run_bazel_output_path(directory)

    external = output_base / "external"

    bazel_workspace = _run_bazel_workspace(directory)

    paths = []
    for action in actions:
        for path in action.inputs + action.outputs:
            if path.suffix in {".cs", ".dll"}:
                for prefix in (
                    external.parent,
                    bazel_workspace,
                    bazel_output_path.parent,
                ):
                    full_path = (prefix / path).resolve()
                    if full_path.exists():
                        paths.append(full_path)

    groupings = {}
    with ThreadPoolExecutor() as executor:
        futures = {
            executor.submit(_list_symbols, Path(list_symbol_binary), path): path
            for path in paths
        }
        for future in concurrent.futures.as_completed(futures):
            symbol = future.result()
            path = futures[future]
            groupings.setdefault(symbol, set()).add(path)

    selected_dlls = set()
    for _, paths in groupings.items():
        if any(path.suffix == ".cs" for path in paths):
            continue
        dll = next((path for path in paths if path.suffix == ".dll"), None)
        if dll:
            selected_dlls.add(dll)

    _generate_csproj(
        csproj_path=directory / f"{project_name}.csproj",
        source_files=[],
        dll_references=list(map(str, selected_dlls)),
        target_framework=target_framework,
    )


def _main():
    argument_parser = argparse.ArgumentParser("Compile C# project")
    argument_parser.add_argument("project_name", type=str, help="Name of the project")
    argument_parser.add_argument(
        "directory", type=Path, help="Path to the project directory"
    )
    argument_parser.add_argument(
        "target_framework", type=str, help="Target framework for the project"
    )
    argument_parser.add_argument(
        "list_symbol_binary", type=Path, help="Path to the ListSymbols binary"
    )
    args = argument_parser.parse_args()
    _compile_csproj(
        args.project_name,
        args.directory,
        args.target_framework,
        args.list_symbol_binary,
    )


if __name__ == "__main__":
    _main()

File: tool/CompileCsproj.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using System.Xml.Linq;
using SymbolAnalysis;

class Action
{
    public int TargetId { get; set; }
    public List<string> Inputs { get; set; } = new List<string>();
    public List<string> Outputs { get; set; } = new List<string>();
    public string PrimaryOutput { get; set; } = "";
}

class ActionsFactory
{
    private readonly JsonDocument _aquery;

    public ActionsFactory(JsonDocument aquery)
    {
        _aquery = aquery;
    }

    public List<Action> GetActions()
    {
        var actions = new List<Action>();
        var aqueryActions = _aquery.RootElement.GetProperty("actions").EnumerateArray();

        foreach (var aqueryAction in aqueryActions)
        {
            if (aqueryAction.GetProperty("mnemonic").GetString() == "CSharpCompile")
            {
                var action = GetAction(aqueryAction.GetProperty("targetId").GetInt32());
                actions.Add(action);
            }
        }
        return actions;
    }

    private string GetArtifactPath(int artifactId)
    {
        var artifacts = _aquery.RootElement.GetProperty("artifacts").EnumerateArray();
        var artifact = artifacts.First(a => a.GetProperty("id").GetInt32() == artifactId);
        var pathFragmentId = artifact.GetProperty("pathFragmentId").GetInt32();
        return GetPath(_aquery.RootElement.GetProperty("pathFragments").EnumerateArray(), pathFragmentId);
    }

    private List<string> RecursivelyGetInputs(int depSetId)
    {
        var inputs = new List<string>();
        var depSetsOfFiles = _aquery.RootElement.GetProperty("depSetOfFiles").EnumerateArray();
        var depSetOfFiles = depSetsOfFiles.First(d => d.GetProperty("id").GetInt32() == depSetId);

        if (depSetOfFiles.TryGetProperty("directArtifactIds", out var directArtifactIds))
        {
            foreach (var directArtifactId in directArtifactIds.EnumerateArray())
            {
                var path = GetArtifactPath(directArtifactId.GetInt32());
                inputs.Add(path);
            }
        }

        if (depSetOfFiles.TryGetProperty("transitiveDepSetIds", out var transitiveDepSetIds))
        {
            foreach (var transitiveDepSetId in transitiveDepSetIds.EnumerateArray())
            {
                inputs.AddRange(RecursivelyGetInputs(transitiveDepSetId.GetInt32()));
            }
        }

        return inputs;
    }

    private Action GetAction(int targetId)
    {
        var inputs = new List<string>();
        var outputs = new List<string>();

        var aqueryActions = _aquery.RootElement.GetProperty("actions").EnumerateArray();
        var action = aqueryActions.First(a =>
            a.GetProperty("targetId").GetInt32() == targetId &&
            a.GetProperty("mnemonic").GetString() == "CSharpCompile");

        if (action.TryGetProperty("inputDepSetIds", out var inputDepSetIds))
        {
            foreach (var inputDepSetId in inputDepSetIds.EnumerateArray())
            {
                var depSetsOfFiles = _aquery.RootElement.GetProperty("depSetOfFiles").EnumerateArray();
                var depSetOfFiles = depSetsOfFiles.First(d => d.GetProperty("id").GetInt32() == inputDepSetId.GetInt32());

                if (depSetOfFiles.TryGetProperty("directArtifactIds", out var directArtifactIds))
                {
                    foreach (var directArtifactId in directArtifactIds.EnumerateArray())
                    {
                        var path = GetArtifactPath(directArtifactId.GetInt32());
                        inputs.Add(path);
                    }
                }

                if (depSetOfFiles.TryGetProperty("transitiveDepSetIds", out var transitiveDepSetIds))
                {
                    foreach (var transitiveDepSetId in transitiveDepSetIds.EnumerateArray())
                    {
                        var paths = RecursivelyGetInputs(transitiveDepSetId.GetInt32());
                        inputs.AddRange(paths);
                    }
                }
            }
        }

        if (action.TryGetProperty("outputIds", out var outputIds))
        {
            foreach (var outputId in outputIds.EnumerateArray())
            {
                var path = GetArtifactPath(outputId.GetInt32());
                outputs.Add(path);
            }
        }

        var primaryOutputId = action.GetProperty("primaryOutputId").GetInt32();
        var primaryOutputPath = GetArtifactPath(primaryOutputId);

        // Filter out DOTNET_CLI_HOME paths
        string dotnetCliHome = "";
        if (action.TryGetProperty("environmentVariables", out var envVars))
        {
            foreach (var envVar in envVars.EnumerateArray())
            {
                if (envVar.GetProperty("key").GetString() == "DOTNET_CLI_HOME")
                {
                    dotnetCliHome = envVar.GetProperty("value").GetString();
                    break;
                }
            }
        }

        if (!string.IsNullOrEmpty(dotnetCliHome))
        {
            inputs = inputs.Where(input => !input.StartsWith(dotnetCliHome)).ToList();
        }

        return new Action
        {
            TargetId = targetId,
            Inputs = inputs,
            Outputs = outputs,
            PrimaryOutput = primaryOutputPath
        };
    }

    private static string GetPath(JsonElement.ArrayEnumerator pathFragments, int pathFragmentId)
    {
        var fragment = pathFragments.First(f => f.GetProperty("id").GetInt32() == pathFragmentId);
        var relatedFragments = new List<string> { fragment.GetProperty("label").GetString()! };

        while (fragment.TryGetProperty("parentId", out var parentIdElement))
        {
            var parentId = parentIdElement.GetInt32();
            fragment = pathFragments.First(f => f.GetProperty("id").GetInt32() == parentId);
            relatedFragments.Add(fragment.GetProperty("label").GetString()!);
        }

        relatedFragments.Reverse();
        return Path.Combine(relatedFragments.ToArray());
    }
}

class CompileCsproj
{
    const string BazelBinary = "bazel";

    static string RunProcess(string[] command, string cwd)
    {
        var psi = new ProcessStartInfo
        {
            FileName = command[0],
            Arguments = string.Join(' ', command.Skip(1)),
            WorkingDirectory = cwd,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false
        };
        using var proc = Process.Start(psi);
        string output = proc.StandardOutput.ReadToEnd();
        string error = proc.StandardError.ReadToEnd();
        proc.WaitForExit();
        if (proc.ExitCode != 0)
            throw new Exception($"Command '{string.Join(' ', command)}' failed.\nStdout: {output}\nStderr: {error}");
        return output;
    }

    static JsonDocument RunBazelAquery(string directory, params string[] targets)
    {
        string queryExpression;

        // Default to all targets if none specified
        if (targets.Length == 0)
        {
            queryExpression = "//...";
        }
        else if (targets.Length == 1)
        {
            queryExpression = targets[0];
        }
        else
        {
            // For multiple targets, use a union expression: "target1 + target2 + target3"
            queryExpression = string.Join(" + ", targets);
        }

        var command = new[] { BazelBinary, "aquery", queryExpression, "--output=jsonproto" };

        string output = RunProcess(command, directory);
        return JsonDocument.Parse(output);
    }

    static string RunBazelInfo(string directory, string info)
    {
        return RunProcess(new[] { BazelBinary, "info", info }, directory).Trim();
    }



    static List<Action> GetActions(string directory, params string[] targets)
    {
        var aquery = RunBazelAquery(directory, targets);
        var actionsFactory = new ActionsFactory(aquery);
        return actionsFactory.GetActions();
    }

    static void CompileCsProj(string projectName, string directory, string targetFramework, params string[] targets)
    {
        string workspaceAbsolute = Environment.GetEnvironmentVariable("BUILD_WORKSPACE_DIRECTORY") ?? "";
        if (!string.IsNullOrEmpty(workspaceAbsolute))
            directory = Path.Combine(workspaceAbsolute, directory);

        var actions = GetActions(directory, targets);
        string outputBase = RunBazelInfo(directory, "output_base");
        string bazelOutputPath = RunBazelInfo(directory, "output_path");
        string bazelWorkspace = RunBazelInfo(directory, "workspace");

        string external = Path.Combine(outputBase, "external");
        var prefixes = new[]
        {
            Path.GetDirectoryName(external)!,
            bazelWorkspace,
            Path.GetDirectoryName(bazelOutputPath)!
        };

        var paths = new List<string>();
        foreach (var action in actions)
        {
            var allPaths = action.Inputs.Concat(action.Outputs);
            foreach (var path in allPaths)
            {
                var extension = Path.GetExtension(path);
                if (extension == ".cs" || extension == ".dll")
                {
                    foreach (var prefix in prefixes)
                    {
                        var fullPath = Path.GetFullPath(Path.Combine(prefix, path));
                        if (File.Exists(fullPath))
                        {
                            paths.Add(fullPath);
                            break;
                        }
                    }
                }
            }
        }

        var groupings = new ConcurrentDictionary<string, ConcurrentBag<string>>();

        Parallel.ForEach(paths, path =>
        {
            try
            {
                var symbol = SymbolLister.AnalyzeFile(path);
                if (symbol != null)
                {
                    groupings.AddOrUpdate(symbol,
                        new ConcurrentBag<string> { path },
                        (key, existing) => { existing.Add(path); return existing; });
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Warning: Failed to get symbols for {path}: {ex.Message}");
            }
        });

        var selectedDlls = new HashSet<string>();
        foreach (var (symbol, pathsBag) in groupings)
        {
            var pathsList = pathsBag.ToList();
            if (pathsList.Any(path => Path.GetExtension(path) == ".cs"))
                continue;

            var dll = pathsList.FirstOrDefault(path => Path.GetExtension(path) == ".dll");
            if (!string.IsNullOrEmpty(dll))
                selectedDlls.Add(dll);
        }

        GenerateCsproj(
            Path.Combine(directory, $"{projectName}.csproj"),
            new List<string>(), // Source files - empty as in Python
            selectedDlls.ToList(),
            targetFramework
        );
    }

    static void GenerateCsproj(string csprojPath, List<string> sourceFiles, List<string> dllReferences, string targetFramework)
    {
        var project = new XElement("Project", new XAttribute("Sdk", "Microsoft.NET.Sdk"));
        var propGroup = new XElement("PropertyGroup");
        propGroup.Add(new XElement("TargetFramework", targetFramework));
        project.Add(propGroup);

        if (sourceFiles.Count > 0)
        {
            var itemGroupSources = new XElement("ItemGroup");
            foreach (var src in sourceFiles)
                itemGroupSources.Add(new XElement("Compile", new XAttribute("Include", src)));
            project.Add(itemGroupSources);
        }

        if (dllReferences.Count > 0)
        {
            var itemGroupRefs = new XElement("ItemGroup");
            foreach (var dll in dllReferences)
            {
                var refElem = new XElement("Reference", new XAttribute("Include", Path.GetFileNameWithoutExtension(dll)));
                refElem.Add(new XElement("HintPath", dll));
                itemGroupRefs.Add(refElem);
            }
            project.Add(itemGroupRefs);
        }

        var doc = new XDocument(new XDeclaration("1.0", "utf-8", "yes"), project);
        doc.Save(csprojPath);
    }

    static void Main(string[] args)
    {
        if (args.Length < 3)
        {
            Console.WriteLine("Usage: CompileCsproj <project_name> <directory> <target_framework> [targets...]");
            Console.WriteLine("If no targets are specified, defaults to '//...'");
            return;
        }
        string projectName = args[0];
        string directory = args[1];
        string targetFramework = args[2];
        string[] targets = args.Skip(3).ToArray(); // Additional targets if provided

        CompileCsProj(projectName, directory, targetFramework, targets);
    }
}